File: react_on_rails_pro/lib/react_on_rails_pro/license_cache.rb
# frozen_string_literal: true

require "json"
require "digest"
require "fileutils"

module ReactOnRailsPro
  # Caches fetched license tokens to disk.
  # Persists across app restarts to reduce API calls.
  # Validates that cached token belongs to the currently configured license_key.
  class LicenseCache
    CACHE_FILENAME = "react_on_rails_pro_license.cache"

    class << self
      def read
        return nil unless cache_path.exist?

        data = JSON.parse(File.read(cache_path))

        return nil unless valid_for_current_key?(data)

        data
      rescue JSON::ParserError, Errno::ENOENT
        nil
      end

      def write(data)
        FileUtils.mkdir_p(cache_dir)

        cache_data = data.merge(
          "license_key_hash" => current_key_hash,
          "fetched_at" => Time.now.iso8601
        )

        File.write(cache_path, JSON.pretty_generate(cache_data))
        File.chmod(0o600, cache_path)
      rescue StandardError => e
        Rails.logger.warn { "[ReactOnRailsPro] Failed to write license cache: #{e.message}" }
      end

      def token
        read&.dig("token")
      end

      def fetched_at
        timestamp = read&.dig("fetched_at")
        Time.parse(timestamp) if timestamp
      rescue ArgumentError
        nil
      end

      def expires_at
        timestamp = read&.dig("expires_at")
        Time.parse(timestamp) if timestamp
      rescue ArgumentError
        nil
      end

      private

      def valid_for_current_key?(data)
        stored_hash = data["license_key_hash"]
        return false if stored_hash.nil?

        stored_hash == current_key_hash
      end

      def current_key_hash
        key = ReactOnRailsPro.configuration.license_key
        return nil if key.nil?

        Digest::SHA256.hexdigest(key)[0..15]
      end

      def cache_dir
        Rails.root.join("tmp")
      end

      def cache_path
        cache_dir.join(CACHE_FILENAME)
      end
    end
  end
end

File: packages/react-on-rails-pro-node-renderer/src/shared/licenseCache.ts
/**
 * Caches fetched license tokens to disk.
 * Persists across app restarts to reduce API calls.
 * Validates that cached token belongs to the currently configured license_key.
 *
 * @module shared/licenseCache
 */

import * as fs from 'fs';
import * as path from 'path';
import * as crypto from 'crypto';

const CACHE_FILENAME = 'react_on_rails_pro_license.cache';

export interface CacheData {
  token: string;
  expires_at: string;
  fetched_at: string;
  license_key_hash: string;
}

/**
 * Gets the current license key from environment.
 */
function getLicenseKey(): string | undefined {
  return process.env.REACT_ON_RAILS_PRO_LICENSE_KEY;
}

/**
 * Computes a hash of the license key for validation.
 * Only stores first 16 chars of SHA256 hash.
 */
function computeKeyHash(key: string): string {
  return crypto.createHash('sha256').update(key).digest('hex').substring(0, 16);
}

/**
 * Gets the current key hash, or null if no key is set.
 */
function getCurrentKeyHash(): string | null {
  const key = getLicenseKey();
  if (!key) {
    return null;
  }
  return computeKeyHash(key);
}

/**
 * Gets the cache directory path.
 * Uses tmp/ relative to current working directory.
 */
function getCacheDir(): string {
  return path.join(process.cwd(), 'tmp');
}

/**
 * Gets the full cache file path.
 */
function getCachePath(): string {
  return path.join(getCacheDir(), CACHE_FILENAME);
}

/**
 * Validates that the cached data belongs to the current license key.
 */
function isValidForCurrentKey(data: CacheData): boolean {
  const storedHash = data.license_key_hash;
  if (!storedHash) {
    return false;
  }

  const currentHash = getCurrentKeyHash();
  return storedHash === currentHash;
}

/**
 * Reads the cached license data from disk.
 * Returns null if cache doesn't exist, is invalid, or belongs to a different license key.
 */
export function readCache(): CacheData | null {
  try {
    const cachePath = getCachePath();

    if (!fs.existsSync(cachePath)) {
      return null;
    }

    const content = fs.readFileSync(cachePath, 'utf8');
    const data = JSON.parse(content) as CacheData;

    if (!isValidForCurrentKey(data)) {
      return null;
    }

    return data;
  } catch {
    return null;
  }
}

/**
 * Writes license data to the cache file.
 * Automatically adds license_key_hash and fetched_at.
 */
export function writeCache(data: { token: string; expires_at: string }): void {
  try {
    const cacheDir = getCacheDir();
    const cachePath = getCachePath();

    // Ensure cache directory exists
    if (!fs.existsSync(cacheDir)) {
      fs.mkdirSync(cacheDir, { recursive: true });
    }

    const currentKeyHash = getCurrentKeyHash();
    if (!currentKeyHash) {
      console.warn('[React on Rails Pro] Cannot write cache: no license key configured');
      return;
    }

    const cacheData: CacheData = {
      token: data.token,
      expires_at: data.expires_at,
      fetched_at: new Date().toISOString(),
      license_key_hash: currentKeyHash,
    };

    fs.writeFileSync(cachePath, JSON.stringify(cacheData, null, 2));

    // Set file permissions to 0600 (owner read/write only)
    fs.chmodSync(cachePath, 0o600);
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : 'Unknown error';
    console.warn(`[React on Rails Pro] Failed to write license cache: ${errorMessage}`);
  }
}

/**
 * Gets the cached token, or null if not available.
 */
export function getCachedToken(): string | null {
  const data = readCache();
  return data?.token ?? null;
}

/**
 * Gets the fetched_at timestamp from cache.
 * Returns null if cache doesn't exist or is invalid.
 */
export function getFetchedAt(): Date | null {
  const data = readCache();
  if (!data?.fetched_at) {
    return null;
  }

  try {
    return new Date(data.fetched_at);
  } catch {
    return null;
  }
}

/**
 * Gets the expires_at timestamp from cache.
 * Returns null if cache doesn't exist or is invalid.
 */
export function getExpiresAt(): Date | null {
  const data = readCache();
  if (!data?.expires_at) {
    return null;
  }

  try {
    return new Date(data.expires_at);
  } catch {
    return null;
  }
}
